### PR TITLE
Bug 1440828 - Phabricator review requests should show up on the BMO dashboard

### PR DESCRIPTION
**Pull request by globau**
Adds a new table to the top of the right hand column that shows reviews from phabricator.

screenshot 2018-03-21 14 05 35

### CODE REVIEW

**Comment by globau** [comment]
note the current test failures are for unfiltered `urlbase` in templates not touched by this PR.

File: extensions/PhabBugz/lib/Util.pm [hunk 1]
```diff
@@ -19,22 +19,23 @@ use Bugzilla::Util qw(trim);
 use Bugzilla::Extension::PhabBugz::Constants;
 
 use JSON::XS qw(encode_json decode_json);
-use List::Util qw(first);
+use List::Util qw(any first);
```
Comment (globau) [inline extensions/PhabBugz/lib/Util.pm]: note to self: no longer using `any`, remove

File: extensions/PhabBugz/lib/Util.pm [hunk 2]
```diff
@@ -536,4 +537,43 @@ sub add_security_sync_comments {
...
+    ThrowCodeError('phabricator_api_error', { reason => 'Malformed Response' })
+        unless exists $diffs->{result}{data};
+
+    # extract this reviewer's status from 'attachments'
+    my @result;
+    foreach my $diff (@{ $diffs->{result}{data} }) {
+        my $attachments = delete $diff->{attachments};
+        foreach my $review (@{ $attachments->{reviewers}{reviewers} }) {
+            next unless $review->{reviewerPHID} eq $phid_user;
+            $diff->{fields}{review_status} = $review->{status};
```
Comment (globau) [inline extensions/PhabBugz/lib/Util.pm]: note to self: skip this diff if `$user` has already reviewed it.  need a list of phab status

**Comment by dylanwh** [comment]
@globau rebase this against master and the urlbase stuff will be fixed.

File: extensions/MyDashboard/template/en/default/pages/mydashboard.html.tmpl [hunk 3]
```diff
@@ -126,29 +127,28 @@
...
-          Flags You Have Requested
-        </div>
-        <span id="requester_count_refresh" class="bz_default_hidden">
-          <span class="items_found" id="requester_flags_found">0 flags found</span>
-          | <a class="refresh" href="javascript:void(0);" id="requester_refresh">Refresh</a>
-          | <a class="buglist" href="javascript:void(0);" id="requester_buglist">Buglist</a>
-        </span>
-        <div id="requester_table"></div>
-      </div>
+      [% ## no-qa
```
Comment (dylanwh) [inline extensions/MyDashboard/template/en/default/pages/mydashboard.html.tmpl]: This is confusing, perhaps call it ## no 008filter?

File: extensions/PhabBugz/lib/WebService.pm [hunk 4]
```diff
@@ -291,6 +299,92 @@ sub obsolete_attachments {
...
+
+    my $reviews = get_needs_review();
+
+    # map author phids to bugzilla users
+    my $author_id_map = get_phab_bmo_ids({
+        phids => [
+            uniq
+            grep { defined }
+            map { $_->{fields}{authorPHID} }
+            @{$reviews}
```
Comment (dylanwh) [inline extensions/PhabBugz/lib/WebService.pm]: nit: curlies not required for de-referencing an array here.
perlcritic is set to allow this, with [-References::ProhibitDoubleSigils].

(in a lot of cases I'll prefer `$foo = $some->{data}{structure}; @$foo`
to `@{ $some->{data}{structure} }` )

File: extensions/PhabBugz/lib/WebService.pm [hunk 5]
```diff
@@ -291,6 +299,92 @@ sub obsolete_attachments {
...
+        @{$reviews}
+    ]);
+
+    # get all bug statuses and summaries in a single query to avoid creation of
+    # many bug objects
+    my %bugs;
+    if (@{$visible_bugs}) {
+        %bugs =
+            map { $_->{bug_id} => $_ }
+            @{ $dbh->selectall_arrayref(
```
Comment (dylanwh) [inline extensions/PhabBugz/lib/WebService.pm]: yeaah, give this expression inside the `@{ }` a name please. :)

File: extensions/PhabBugz/lib/WebService.pm [hunk 6]
```diff
@@ -291,6 +299,92 @@ sub obsolete_attachments {
...
+            uniq
+            grep { defined }
+            map { $_->{fields}{authorPHID} }
+            @{$reviews}
+        ]
+    });
+    my %author_phab_to_id = map { $_->{phid} => $_->{id} } @{$author_id_map};
+    my %author_id_to_user =
+        map { $_->id => $_ }
+        @{ Bugzilla::User->new_from_list([ map { $_->{id} } @{$author_id_map} ]) };
```
Comment (dylanwh) [inline extensions/PhabBugz/lib/WebService.pm]:
```perl
 my $bugzilla_users = Bugzilla::User->new_from_list([ map { $_->{id} } @$author_id_map ])
 my %author_id_to_user = map { $_->id => $_ } @$bugzilla_users;
```

File: extensions/PhabBugz/lib/WebService.pm (hunk 4 above)
Comment (globau) [reply extensions/PhabBugz/lib/WebService.pm]: will do – all my personal stuff has `References::ProhibitDoubleSigils` enabled, so it's habitual

File: extensions/PhabBugz/lib/Util.pm [hunk 7]
```diff
@@ -536,4 +537,41 @@ sub add_security_sync_comments {
...
+        'differential.revision.search',
+        {
+            attachments => {
+                reviewers => 1,
+            },
+            constraints => {
+                reviewerPHIDs => [$phid_user],
+                statuses      => [qw( needs-review )],
+            },
+            order       => 'newest',
```
Comment (purelogiq) [inline extensions/PhabBugz/lib/Util.pm]: Phab's api's default limit is 100 on searches. I'm not sure how fast/slow the dashboard is so we could limit that further here if needed. Hopefully, no one lets their review queue get that big!

Comment (globau) [reply extensions/PhabBugz/lib/Util.pm]: 100 rows isn't a problem for the dashboard performance-wise, but it's a reasonable cutoff for this table.

**Comment by globau** [comment]
you can do some testing locally by setting `$phid_user` in `PhabBugz::Util::get_needs_review()` to the phid of requestee.  the author in the responses are unlikely to be correct due to a missing phid-->bmo mapping, and will show up as "anonymous".

**Comment by dklawren** [comment]
On the latest changes, I get the following error from the dashboard:

Insecure dependency in parameter 1 of DBI::db=HASH(0x55c37cc4a2a0)->selectall_arrayref method call while running with -T switch at /app/extensions/PhabBugz/lib/WebService.pm line 336.

dkl

**Comment by dylanwh** [comment]
I'll look into the error dkl saw and propose a fix for when glob gets back.

**Comment by globau** [comment]
@dylanwh leave it with me; i'll look at this once i've cleared the backlog caused by last week

**Comment by globau** [comment]
@dklawren what steps are you performing to trigger the taint error?  i've been unable to reproduce it within the `bmo-web.vm` vagrant box.  i can confirm that at that point of the code taint mode is enabled (as expected), however `$reviews` isn't tainted, so as a result `$visible_bugs` isn't either.

**Comment by globau** [comment]
that said, i'd forgotten that `$dbh->sql_in()` doesn't automatically quote (which imho smells like a footgun).  i'll fix that up and push the updates.

**Comment by dylanwh** [comment]
- fixup this branch by rebasing it into a single commit
- merge that into the development branch, which will trigger a new docker image: mozillabteam/bmo:build-NNNN
- ask ckolos to deploy that to bugzilla-dev (and only bugzilla-dev)

**Comment by zalun** [comment]
I've hit the same error as @dklawren 
I've pulled the @globau branch to `bmo` repo and run from `demo` repo with `docker-compose -f docker-compose.yml -f docker-compose.bmo.yml up --build`

**Comment by globau** [comment]
for reasons that aren't clear to me, data returned from phabricator was marked as tainted in the bmo docker container, but not in the bmo-web development virtual machine.  b6905ec detaints all phabricator data.

**Comment by dylanwh** [comment]
Filed [Bug 1453370 - bmo-build-7805](https://bugzilla.mozilla.org/show_bug.cgi?id=1453370) for deploying this change to bugzilla-dev.

**Comment by dylanwh** [comment]
I left this in a bit of an unclear state: This is on bugzilla-dev,
it should be tested by someone that has reviews on phab-dev.

After that, r+ from me.

**Pull request merged**